app/src/main/java/org/wikipedia/login/LoginActivity.kt

```diff
@@ -0,0 +1,265 @@
+    private var shouldLogLogin = true
+    private lateinit var binding: ActivityLoginBinding
+    public override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        binding = ActivityLoginBinding.inflate(layoutInflater)
+        binding.viewLoginError.backClickListener = View.OnClickListener { onBackPressed() }
+        binding.viewLoginError.retryClickListener = View.OnClickListener { binding.viewLoginError.visibility = View.GONE }
+
+        // Don't allow user to attempt login until they've put in a username and password
+        NonEmptyValidator(binding.loginButton, binding.loginUsernameText, binding.loginPasswordInput)
+        binding.loginPasswordInput.editText?.setOnEditorActionListener { _, actionId, _ ->
+            if (actionId == EditorInfo.IME_ACTION_DONE) {
+                validateThenLogin()
+                return@setOnEditorActionListener true
+            }
+            false
+        }
+        funnel = LoginFunnel(WikipediaApp.getInstance())
+        loginSource = intent.getStringExtra(LOGIN_REQUEST_SOURCE)
+        if (!TextUtils.isEmpty(loginSource) && loginSource == LoginFunnel.SOURCE_SUGGESTED_EDITS) {
+            Prefs.setSuggestedEditsHighestPriorityEnabled(true)
+        }
+
+        // always go to account creation before logging in
+        if (savedInstanceState == null) {
+            startCreateAccountActivity()
+        }
+        setAllViewsClickListener()
+
+        // Assume no login by default
+        setResult(RESULT_LOGIN_FAIL)
+    }
+
+    private fun setAllViewsClickListener() {
+        binding.loginButton.setOnClickListener { validateThenLogin() }
+        binding.loginCreateAccountButton.setOnClickListener { startCreateAccountActivity() }
+        binding.inflateLoginAndAccount.privacyPolicyLink.setOnClickListener { showPrivacyPolicy(this) }
+        binding.inflateLoginAndAccount.forgotPasswordLink.setOnClickListener {
+            val title = PageTitle("Special:PasswordReset", WikipediaApp.getInstance().wikiSite)
+            visitInExternalBrowser(this, Uri.parse(title.uri))
+        }
+    }
+
+    private fun getText(input: TextInputLayout): String {
+        return StringUtils.defaultString(if (input.editText != null && input.editText!!.text != null) input.editText!!.text.toString() else "")
+    }
+
+    private fun clearErrors() {
+        binding.loginUsernameText.isErrorEnabled = false
+        binding.loginPasswordInput.isErrorEnabled = false
+    }
+
+    private fun validateThenLogin() {
+        clearErrors()
+        if (!CreateAccountActivity.USERNAME_PATTERN.matcher(getText(binding.loginUsernameText)).matches()) {
+            binding.loginUsernameText.requestFocus()
+            binding.loginUsernameText.error = getString(R.string.create_account_username_error)
+            return
+        }
+        doLogin()
+    }
+
+    private fun logLoginStart() {
+        if (shouldLogLogin && !TextUtils.isEmpty(loginSource)) {
```

Comment:
All instances of `TextUtils.isEmpty()` can be replaced with kotlin's built-in `[string].isNullOrEmpty()`.